addons/edi/models/edi_document.py

```diff
@@ -367,6 +393,52 @@ def commit_processing_state_change(self, processing_state=""):
             self.processing_state = processing_state
             self.env.cr.commit()
 
+    def create_or_update_edi_progress(self, log_message=None):
+        """EDI records processing is triggered in `prepare()` and `execute()` which creates a cursor
+        savepoint to atomically commit/rollback the created/updated records. To report the progress
+        of long running executions, it is required to continually update the so-far-processed status
+        log, **while** the records are being processed. As the EDI document is locked at row-level
+        before starting process (in `lock_for_action`), writing to that record will deadlock.
+        Instead, an instance of EdiDocumentProgress will keep the progress count (which is linked to
+        an EDIDocument). The progress update logs must be committed to the DB to display the
+        progress in UI/view.
+
+        In order to commit change to 'another object' (EdiDocumentProgress in this case), create
+        another DB cursor. Transaction is committed on closing the cursor. This leaves the
+        encompassing transaction (EDI records processing) undisturbed."""
+        self.ensure_one()
+        towrite = self.env.all.towrite.copy()
+        tocompute = self.env.all.tocompute.copy()
+        with self.pool.cursor() as temp_cr:
+            # NOTE: this method does a commit using another cursor, but it is called from inside a
+            # savepoint. Odoo core has a bug that it commits all pending writes without considering
+            # the cursor. To work around, clear the pending writes & computes here; write the progress
+            # and commit; then restore those after.
+            temp_self = self.with_env(self.env(cr=temp_cr))
+            temp_self.env.all.towrite.clear()
```

Comment:
Is there a race condition here? For example, what if some other cursor from the pool calls `commit()` while the dicts are empty? (I suppose the first commit after the dicts are restored would pick up the skipped updates, but ... :grimacing: 

Or have I missed something obvious due to lack of morning caffeine?

Reply:
It seems that dicts are shared across the request's env cursors rather than the whole pool, so I'll resolve this.